File: frontend/packages/core/src/types/provider.ts
export type WireApi = 'openai-completions' | 'openai-responses' | 'anthropic-messages'

export type Readiness =
  'ready' | 'degraded' | 'stale' | 'provider_error' | 'auth_error' | 'model_error' | 'unavailable'

export interface AuthSpec {
  mode: 'api_key' | 'bearer' | 'none' | 'oauth' | 'iam'
  header_name?: string
  header_prefix?: string
}

export interface EndpointPreset {
  preset_key: string
  region: string
  protocol: WireApi
  plan: string | null
  base_url: string
  model_ids: string[]
  /** Resolved capability descriptor for this endpoint. The wizard prefills the
   *  capability editor with it and sends it back only when the user overrides it. */
  capability: Record<string, unknown>
}

export interface ModelPresetEntry {
  model_id: string
  display_name: string
  plan: string[] | null
  context_window: number
  max_tokens: number
  input_modalities: string[]
  reasoning: boolean
  pricing: { input: number; output: number; cache_read?: number; cache_write?: number }
}

export interface VendorPreset {
  vendor: string
  display_name: string
  short_name: string
  logo: string | null
  category: 'saas' | 'oss-framework' | 'custom'
  description: string
  endpoints: EndpointPreset[]
  models: ModelPresetEntry[]
}

export interface ProbeStep {
  name: string
  status: 'pass' | 'fail' | 'skip' | 'warn'
  latency_ms?: number | null
  detail?: string
  error?: { type: string; message: string; raw_status?: number | null } | null
}

export interface ProbeResult {
  overall: 'pass' | 'fail' | 'warn' | 'unavailable'
  blocking_failed: boolean
  steps: ProbeStep[]
}

export interface Provider {
  id: string
  name: string
  slug: string
  provider_type: string
  base_url: string
  auth_type: 'api_key' | 'oauth' | 'bearer_token' | 'none'
  has_api_key: boolean
  logo_url: string | null
  enabled: boolean
  is_system: boolean
  model_count: number
  models?: Model[]
  extra_body: Record<string, unknown>
  extra_headers: Record<string, unknown>
  created_by_user_id: string
  created_at: string
  updated_at: string
  preset_slug?: string | null
  logo?: string | null
  capability?: Record<string, unknown>
  model_capability_overrides?: Record<string, unknown>
  last_liveness_status?: string | null
  last_liveness_at?: string | null
}

export interface Model {
  id: string
  provider_id: string
  model_id: string
  display_name: string
  reasoning: boolean
  input_modalities: string[]
  cost_input: number
  cost_output: number
  cost_cache_read: number
  cost_cache_write: number
  context_window: number
  max_tokens: number
  extra_body: Record<string, unknown>
  extra_headers: Record<string, unknown>
  enabled: boolean
  is_system: boolean
  created_at: string
  updated_at: string
  last_test_status?: string
  last_test_at?: string | null
  last_test_summary?: Record<string, unknown>
  readiness?: Readiness
}

export interface ProviderCreate {
  name: string
  slug?: string
  provider_type?: string
  base_url: string
  auth_type?: string
  api_key?: string | null
  logo_url?: string | null
  extra_body?: Record<string, unknown>
  extra_headers?: Record<string, unknown>
  preset_slug?: string
  capability?: Record<string, unknown>
  model_capability_overrides?: Record<string, unknown>
}

export interface ProviderUpdate {
  name?: string | null
  provider_type?: string | null
  base_url?: string | null
  auth_type?: string | null
  api_key?: string | null
  logo_url?: string | null
  extra_body?: Record<string, unknown> | null
  extra_headers?: Record<string, unknown> | null
  enabled?: boolean | null
  preset_slug?: string
  capability?: Record<string, unknown>
  model_capability_overrides?: Record<string, unknown>
}

export interface ModelCreate {
  model_id: string
  display_name: string
  reasoning?: boolean
  input_modalities?: string[]
  cost_input?: number
  cost_output?: number
  cost_cache_read?: number
  cost_cache_write?: number
  context_window: number
  max_tokens: number
  extra_body?: Record<string, unknown>
  extra_headers?: Record<string, unknown>
  enabled?: boolean
}

export interface ModelUpdate {
  display_name?: string | null
  reasoning?: boolean | null
  input_modalities?: string[] | null
  cost_input?: number | null
  cost_output?: number | null
  cost_cache_read?: number | null
  cost_cache_write?: number | null
  context_window?: number | null
  max_tokens?: number | null
  extra_body?: Record<string, unknown> | null
  extra_headers?: Record<string, unknown> | null
  enabled?: boolean | null
}

File: backend/cubebox/api/schemas/provider.py
"""Request/response schemas for provider & model admin API."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

WireApi = Literal["openai-completions", "openai-responses", "anthropic-messages"]


class ProviderCreate(BaseModel):
    name: str = Field(max_length=64)
    slug: str | None = Field(default=None, max_length=64)
    provider_type: WireApi = "openai-completions"
    base_url: str = Field(max_length=2048)
    auth_type: str = Field(default="api_key", max_length=32)
    api_key: str | None = Field(default=None, max_length=512)
    logo_url: str | None = Field(default=None, max_length=512)
    extra_body: dict[str, Any] = Field(default_factory=dict)
    extra_headers: dict[str, Any] = Field(default_factory=dict)
    preset_slug: str | None = Field(default=None, max_length=64)
    capability: dict[str, Any] = Field(default_factory=dict)
    model_capability_overrides: dict[str, Any] = Field(default_factory=dict)


class ProviderUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=64)
    provider_type: WireApi | None = None
    base_url: str | None = Field(default=None, max_length=2048)
    auth_type: str | None = Field(default=None, max_length=32)
    api_key: str | None = Field(default=None, max_length=512)
    logo_url: str | None = Field(default=None, max_length=512)
    extra_body: dict[str, Any] | None = None
    extra_headers: dict[str, Any] | None = None
    enabled: bool | None = None
    preset_slug: str | None = None
    capability: dict[str, Any] | None = None
    model_capability_overrides: dict[str, Any] | None = None


class ModelTest(BaseModel):
    model_id: str = Field(max_length=128)


class ProviderLivenessRequest(BaseModel):
    """Pre-save liveness dry-run body (spec §4.3).

    Builds a transient provider from these fields — no row is written. ``model_id``
    is the cheap model the single liveness call is issued against.
    """

    preset_slug: str | None = Field(default=None, max_length=64)
    api: WireApi = "openai-completions"
    base_url: str = Field(max_length=2048)
    api_key: str | None = Field(default=None, max_length=512)
    capability: dict[str, Any] = Field(default_factory=dict)
    model_capability_overrides: dict[str, Any] = Field(default_factory=dict)
    model_id: str = Field(max_length=128)


class ProviderTestRequest(ProviderLivenessRequest):
    """Pre-save full-probe dry-run body (spec §4.3).

    Same transient-provider construction as liveness, but runs liveness + the
    per-model capability probe and returns the composed ``ProbeResult``.
    """


class ProviderTestStreamRequest(BaseModel):
    """Explicit model DB ids to test (wizard models are enabled=false)."""

    model_db_ids: list[str] = Field(min_length=1)


class OrgProviderOverrideUpdate(BaseModel):
    enabled: bool


class OrgProviderOverrideOut(BaseModel):
    enabled: bool


class ModelCreate(BaseModel):
    model_id: str = Field(max_length=128)
    display_name: str = Field(max_length=128)
    reasoning: bool = False
    input_modalities: list[str] = Field(default_factory=lambda: ["text"])
    cost_input: float = 0.0
    cost_output: float = 0.0
    cost_cache_read: float = 0.0
    cost_cache_write: float = 0.0
    context_window: int
    max_tokens: int
    extra_body: dict[str, Any] = Field(default_factory=dict)
    extra_headers: dict[str, Any] = Field(default_factory=dict)
    enabled: bool = True


class ModelUpdate(BaseModel):
    display_name: str | None = Field(default=None, max_length=128)
    reasoning: bool | None = None
    input_modalities: list[str] | None = None
    cost_input: float | None = None
    cost_output: float | None = None
    cost_cache_read: float | None = None
    cost_cache_write: float | None = None
    context_window: int | None = None
    max_tokens: int | None = None
    extra_body: dict[str, Any] | None = None
    extra_headers: dict[str, Any] | None = None
    enabled: bool | None = None


class ModelOut(BaseModel):
    id: str
    provider_id: str
    model_id: str
    display_name: str
    reasoning: bool
    input_modalities: list[str]
    cost_input: float
    cost_output: float
    cost_cache_read: float
    cost_cache_write: float
    context_window: int
    max_tokens: int
    extra_body: dict[str, Any]
    extra_headers: dict[str, Any]
    enabled: bool
    is_system: bool
    created_at: datetime
    updated_at: datetime


class ModelReadinessOut(ModelOut):
    """A model row plus its per-model test status and server-derived readiness.

    Returned by GET /admin/providers/{id} so the UI reads `readiness` verbatim
    (it never re-derives). `last_test_at` is a UTC-offset ISO string.
    """

    last_test_at: str | None = None
    last_test_status: str | None = None  # "ok" | "warn" | "fail" | "unavailable"
    last_test_summary: dict[str, Any] = Field(default_factory=dict)
    # §4.1: "ready"|"degraded"|"stale"|"model_error"|"unavailable"|"provider_error"
    readiness: str


class ProviderOut(BaseModel):
    id: str
    name: str
    slug: str
    # Read-side: reflects the DB column (a plain str). The 3-literal contract is
    # enforced on the write path (ProviderCreate/ProviderUpdate), so stored values
    # are already canonical; no need to re-validate on every read.
    provider_type: str
    base_url: str
    auth_type: str
    has_api_key: bool
    logo_url: str | None
    enabled: bool
    is_system: bool
    model_count: int
    models: list[ModelReadinessOut] | None = None
    org_override: OrgProviderOverrideOut | None = None
    extra_body: dict[str, Any]
    extra_headers: dict[str, Any]
    preset_slug: str | None = None
    # Brand-icon id resolved from the preset catalog so the configured-only
    # provider UI can render the brand glyph without fetching presets.
    logo: str | None = None
    capability: dict[str, Any] = Field(default_factory=dict)
    model_capability_overrides: dict[str, Any] = Field(default_factory=dict)
    # Provider-level liveness/credential status (spec §4.1). UTC-offset ISO string.
    last_liveness_at: str | None = None
    last_liveness_status: str | None = None  # "ok" | "fail"
    last_liveness_summary: dict[str, Any] = Field(default_factory=dict)
    created_by_user_id: str | None
    created_at: datetime
    updated_at: datetime
